php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +64,349 @@ class StudentCompetency extends \ActiveRecord
+                REPLACE(
+                   IFNULL (
+                       JSON_EXTRACT(DemonstrationsRequired, \'$."%u"\'),
+                       JSON_EXTRACT(DemonstrationsRequired, "$.default")
+                   ),
+                   \'"\',
+                   ""
+                ) AS DemonstrationsRequirements
+              FROM `%s`
+             WHERE CompetencyID = %u
+            HAVING DemonstrationsRequirements > 0
+            ',
+            [
+                $this->Level,
+                Skill::$tableName,
+                $this->Competency->ID
+            ]
+        );
+        $validSkillIds = array_column($validSkills, 'ID');
+
+        $ratedSkills = DB::valuesTable(
+            'SkillID',
+            'skillRatings',
+            '
+            SELECT COUNT(*) AS skillRatings, SkillID
+        	  FROM `%1$s` ds
+    		  JOIN `%2$s` d
+    			ON d.ID = ds.DemonstrationID
+    		 WHERE ds.SkillID IN (%3$s)
+    		   AND ds.TargetLevel = %4$u
+    		   AND ds.DemonstratedLevel > 0
+    		   AND d.StudentID = %5$u
+    		   AND ds.Override = false
+    		 GROUP BY SkillID
+            ',
+            [
+                DemonstrationSkill::$tableName,
+                Demonstration::$tableName,
+                join(', ', $validSkillIds),
+                $this->Level,
+                $this->StudentID
+            ]
+        );
+
+        // starting rating can not be calculated if each required skill has not been rated at least once
+        foreach ($validSkillIds as $skillId) {
+            if (empty($ratedSkills[$skillId])) {
+                return null;
+            }
+        }
+
+        // get the first rating (by date) for each skill
+        $demonstrationSkillRatings = DB::allValues(
+            'skillRatings',
+            '
+            SELECT ds.DemonstratedLevel as skillRatings
+        	  FROM `%1$s` ds
+    		  JOIN `%2$s` d
+    		    ON d.ID = ds.DemonstrationID
+    		 INNER JOIN (
+                    SELECT MIN(demo.Demonstrated) as Created, SkillID
+    		          FROM `%1$s` demoskill
+    		          JOIN `%2$s` demo
+    		            ON demo.ID = demoskill.DemonstrationID
+    		         WHERE TargetLevel = %3$u
+                       AND DemonstratedLevel > 0
+                       AND Override = false
+                       AND demoskill.SkillID IN (%4$s)
+                       AND demo.StudentID = %5$u
+                     GROUP BY SkillID
+    		  ) ds2
+    		    ON ds2.SkillID = ds.SkillID
+    		   AND ds2.Created = d.Demonstrated
+    		 WHERE ds.SkillID IN (%4$s)
+    		   AND ds.TargetLevel = %3$u
+    		   AND ds.DemonstratedLevel > 0
+    		   AND d.StudentID = %5$u
+    		   AND ds.Override = false
+            ',
+            [
+                DemonstrationSkill::$tableName,
+                Demonstration::$tableName,
+                $this->Level,
+                join(', ', $validSkillIds),
+                $this->StudentID
+            ]
+        );
```

Comment:
@nbey was there any issue with the code I developed here: https://github.com/SlateFoundation/slate-cbl/pull/383#discussion_r132990589

that got us everything we needed in one query